File: leetcode/Python/insert-delete-getrandom-o1.py
class RandomizedSet:

    def __init__(self):
        self.um = dict()
        self.v = list()
        self.idx = 0
    def insert(self, val: int) -> bool:
        if val in self.um:
            return False
        self.um[val]= len(self.v)
        self.v.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.um:
            return False
        idx = self.um[val]
        self.v[idx], self.um[self.v[-1]] = self.v[-1], idx
        self.v.pop()
        del self.um[val]
        return True

    def getRandom(self) -> int:
        return random.choice(self.v)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: leetcode/C++/insert-delete-getrandom-o1.cpp
class RandomizedSet {
public:
    vector<int> v;
    unordered_map<int,int> um;
    int t,a_idx,b_idx;
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if(um.find(val)!=um.end()) return false;
        um[val]=v.size();
        v.push_back(val);
        return true;
    }
    
    bool remove(int val) {
        if(um.find(val)==um.end()) return false;
        a_idx = um[val];
        b_idx = (int)v.size()-1;
        um[v[b_idx]]=a_idx;
        v[a_idx]=v[b_idx];
        v.pop_back();
        um.erase(val);
        return true;
    }
    
    int getRandom() {
        return v[rand() % v.size()];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */